extmod/os_dupterm.c

```diff
@@ -148,29 +148,41 @@ int mp_os_dupterm_rx_chr(void) {
     return -1;
 }
 
-void mp_os_dupterm_tx_strn(const char *str, size_t len) {
+int mp_os_dupterm_tx_strn(const char *str, size_t len) {
+    // Returns the minimum successful write length, or -1 if no write is attempted.
+    int ret = len, did_write = 0;
     for (size_t idx = 0; idx < MICROPY_PY_OS_DUPTERM; ++idx) {
         if (MP_STATE_VM(dupterm_objs[idx]) == MP_OBJ_NULL) {
             continue;
         }
+        did_write = 1;
 
         #if MICROPY_PY_OS_DUPTERM_BUILTIN_STREAM
         if (mp_os_dupterm_is_builtin_stream(MP_STATE_VM(dupterm_objs[idx]))) {
             int errcode = 0;
             const mp_stream_p_t *stream_p = mp_get_stream(MP_STATE_VM(dupterm_objs[idx]));
-            stream_p->write(MP_STATE_VM(dupterm_objs[idx]), str, len, &errcode);
+            int write_res = MAX(0, stream_p->write(MP_STATE_VM(dupterm_objs[idx]), str, len, &errcode));
```

Comment:
This may call the stream write function twice due to MAX being a macro.  So please put the write in a separate statement.

Reply:
good point